File: NeetCodeIO/84LargestRectangleInHistogram/LargestRectangleInHistogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # Stack -> Time:O(n), Space:O(1)
        ans = 0
        stack = []
        for i, h in enumerate(heights):
            start = i
            while stack and stack[-1][1] > h:
                idx, ht = stack.pop()
                ans = max(ans, ht * (i - idx))
                start = idx
            stack.append((start, h))
        for i, h in stack:
            ans = max(ans, h*(len(heights)-i))
        return ans

        # # Brute Force -> Time:O(n^2), Space:O(1) -> Time Limit Exceeded
        # ans = 0
        # for i in range(len(heights)):
        #     m = heights[i]
        #     if m > ans:
        #         ans = m
        #     for j in range(i+1, len(heights)):
        #         if heights[j] < m:
        #             m = heights[j]
        #         # print(i, j, m)
        #         if m * (j-i+1) > ans:
        #             ans = m * (j-i+1)
        # return ans

File: NeetCodeIO/84LargestRectangleInHistogram/LargestRectangleInHistogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        // Stack -> Time:O(n), Space:O(1)
        stack<pair<int,int>> s;
        int area, start, ans = 0;
        pair<int,int> tempPair;
        for (int i = 0; i < heights.size(); i++) {
            start = i;
            while (!s.empty() && s.top().second > heights[i]) {
                tempPair = s.top();
                s.pop();
                ans = max(ans, tempPair.second * (i - tempPair.first));
                start = tempPair.first;
            }
            tempPair.first = start;
            tempPair.second = heights[i];
            s.push(tempPair);
        }
        while (!s.empty()) {
            tempPair = s.top();
            s.pop();
            area = tempPair.second * (heights.size() - tempPair.first);
            ans = max(ans, area);
        }
        return ans;
    }
};
